src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2024,147 +1981,192 @@ else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP)) {
+                while ((runState & STOP) == 0L && ((p = w.phase) & IDLE) != 0) {
+                    int trim;
+                    if (!Thread.interrupted())
+                        U.park(trimmable, deadline);
+                    if (trimmable) {
+                        if ((trim = tryTrim(w, queuedCtl, deadline)) > 0)
                             break;
-                        Thread.interrupted();    // clear for next park
-                        if (deadline != 0L && TIMEOUT_SLOP >
-                            deadline - System.currentTimeMillis()) {
-                            long sp = w.stackPred & LMASK, c = ctl;
-                            long nc = sp | (UMASK & (c - TC_UNIT));
-                            if (((int)c & SMASK) == (active & SMASK) &&
-                                compareAndSetCtl(c, nc)) {
-                                w.source = DEREGISTERED;
-                                w.phase = active;
-                                break;           // trimmed on timeout
-                            }
-                            deadline = 0L;       // no longer trimmable
+                        else if (trim < 0) {
+                            trimmable = false;
+                            deadline = 0L;
                         }
                     }
-                    w.parking = 0;
-                    LockSupport.setCurrentBlocker(null);
                 }
+                w.source = 0;
+                LockSupport.setCurrentBlocker(null);
             }
         }
         return p;
     }
 
+    /**
+     * Tries to remove and deregister worker after timeout, and release
+     * another to do the same.
+     * @return > 0: trimmed, < 0 : not trimmable, else 0
+     */
+    private int tryTrim(WorkQueue w, long queuedCtl, long deadline) {
+        long c, nc; int stat, p, vp, i; WorkQueue[] vs; WorkQueue v;
+        if ((c = ctl) != queuedCtl || w == null || ((p = w.phase) & IDLE) == 0)
+            stat = -1;
+        else if (deadline - System.currentTimeMillis() >= TIMEOUT_SLOP)
+            stat = 0;                       // spurious wakeup
+        else if (!compareAndSetCtl(
+                     c, nc = (w.stackPred & LMASK) | (UMASK & (c - TC_UNIT))))
+            stat = -1;                      // lost race to signaller
+        else {
+            stat = 1;
+            w.source = DROPPED;
+            w.phase = p + IDLE;
+            if ((vp = (int)nc) != 0 && (vs = queues) != null &&
+                vs.length > (i = vp & SMASK) && (v = vs[i]) != null &&
+                compareAndSetCtl(           // try to wake up next waiter
+                    nc, ((UMASK & (nc + RC_UNIT)) |
+                         (nc & TC_MASK) | (v.stackPred & LMASK)))) {
+                v.source = INVALID_ID;      // enable cascaded timeouts
+                v.phase = vp;
+                U.unpark(v.owner);
+            }
+        }
+        return stat;
```

Comment:
@DougLea Since `stat` is a local, and is only written to once per branch it might make sense to just return from each of the branches?

Reply:
A similar answer as above: using a single return avoids compiler generating branches to its own synthetic single return. Worthwhile only when whole method is of this form (I did check this one.)

Reply:
I prefer having methods with a single return, placed on the last line of the method, but I wanted to make sure I wasn't missing any underlying reason. :)